File: app/src/main/java/com/mdkashif/alarm/activities/SettingsActivity.java
package com.mdkashif.alarm.activities;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.preference.RingtonePreference;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.Theme;
import com.mdkashif.alarm.R;
import com.mdkashif.alarm.utils.AppConstants;
import com.mdkashif.alarm.utils.SharedPrefHolder;
import com.pkmmte.view.CircularImageView;

import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import jp.wasabeef.blurry.Blurry;

public class SettingsActivity extends AppCompatActivity {


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(getResources().getString(R.string.settings));
        getFragmentManager().beginTransaction().replace(android.R.id.content, new MainPreferenceFragment()).commit();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    public static class MainPreferenceFragment extends PreferenceFragment {
        @Override
        public void onCreate(final Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.activity_settings);
            Preference toggleAboutDev= findPreference(getString(R.string.aboutDev));
            Preference toggleRate = findPreference(getString(R.string.rate));
            Preference toggleShare = findPreference(getString(R.string.share));
            Preference toggleAppVersion = findPreference(getString(R.string.keyAppVersion));
            Preference toggleFaq = findPreference(getString(R.string.title_faq));
            Preference togglePrivacyPolicy = findPreference(getString(R.string.privacy_policy));
            Preference toggleTnc = findPreference(getString(R.string.title_terms));
            Preference toggleFeedback = findPreference(getString(R.string.key_send_feedback));
            Preference toggleNotification=findPreference(getString(R.string.notifications_new_message));
            Preference toggleVibrate=findPreference(getString(R.string.key_vibrate));
            final Preference toggleTheme=findPreference(getString(R.string.theme));

            toggleTheme.setSummary(SharedPrefHolder.getInstance(getActivity()).getTheme());
            toggleAboutDev.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    MaterialDialog aboutDev=new MaterialDialog.Builder(getActivity())
                            .theme(Theme.LIGHT)
                            .cancelable(true)
                            .title(getString(R.string.aboutDevTitle))
                            .customView(R.layout.layout_about_dev,false)
                            .show();

                    View view = aboutDev.getCustomView();

                    CircularImageView avatar;
                    Bitmap bitmap;
                    ImageView ivDevBlurryImage,ivLinkedIn,ivGmail,ivGooglePlay,ivStackOverFlow,ivGithub;
                    TextView tvMadeWithLove,tvOpenLicenses, tvVersion;

                    avatar = view.findViewById(R.id.avatar);
                    ivLinkedIn=view.findViewById(R.id.linkedin);
                    ivGmail=view.findViewById(R.id.gmail);
                    ivGooglePlay=view.findViewById(R.id.googleplay);
                    tvOpenLicenses=view.findViewById(R.id.opensource);
                    ivStackOverFlow=view.findViewById(R.id.stackOverFlow);
                    ivGithub=view.findViewById(R.id.github);
                    ivDevBlurryImage=view.findViewById(R.id.ivDevBlurryImage);
                    tvMadeWithLove=view.findViewById(R.id.madeWithLove);
                    tvVersion=view.findViewById(R.id.tvVersion);

                    Spannable spannable = new SpannableString(getString(R.string.summary_about));
                    spannable.setSpan(new ForegroundColorSpan(getResources().getColor(R.color.red)), 10, 11, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                    tvMadeWithLove.setText(spannable, TextView.BufferType.SPANNABLE);

                    if(avatar.getDrawable()!=null) {
                        bitmap = ((BitmapDrawable) avatar.getDrawable()).getBitmap();
                        Blurry.with(getActivity())
                                .radius(5)
                                .sampling(2)
                                .color(Color.argb(100, 0, 0, 0))
                                .async()
                                .animate(1000).from(bitmap).into(ivDevBlurryImage);
                    }

                    ivLinkedIn.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("linkedin://add/%@" + "mdkashif2093"));
                            final PackageManager packageManager = getActivity().getPackageManager();
                            final List<ResolveInfo> list = packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
                            if (list.isEmpty()) {
                                intent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://www.linkedin.com/profile/view?id=" + "mdkashif2093"));
                            }
                            startActivity(intent);
                        }
                    });

                    ivGmail.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {

                            String addresses="mohammadkshf2093@gmail.com";
                            Uri uri = Uri.parse("mailto:" + addresses).buildUpon().build();
                            Intent emailIntent = new Intent(Intent.ACTION_SENDTO, uri);
                            emailIntent.putExtra(Intent.EXTRA_SUBJECT, "Query!");
                            getActivity().startActivity(emailIntent);

                        }
                    });

                    ivGooglePlay.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            try {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://search?q=pub:Mohammad+Kashif+Khan")));
                            } catch (android.content.ActivityNotFoundException anfe) {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/developer?id=Mohammad+Kashif+Khan")));
                            }
                        }
                    });

                    ivStackOverFlow.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://stackoverflow.com/users/5518744/kashif-k")));
                        }
                    });

                    ivGithub.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/mohammadkashifkhan")));
                        }
                    });

                    tvOpenLicenses.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            new MaterialDialog.Builder(getActivity())
                                    .theme(Theme.LIGHT)
                                    .cancelable(true)
                                    .title(getString(R.string.openSourceLicenses))
                                    .customView(R.layout.layout_open_source_licenses,true)
                                    .show();
                        }
                    });

                    try {
                        PackageInfo pInfo = getActivity().getPackageManager().getPackageInfo(getActivity().getPackageName(), 0);
                        tvVersion.setText(pInfo.versionName);
                    }
                    catch (Exception e)
                    {
                        e.printStackTrace();
                    }

                    return true;
                }
            });

            toggleNotification.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object o) {

                    if(Boolean.valueOf(o.toString()))
                        SharedPrefHolder.getInstance(getActivity()).setRingStatus(true);
                    else
                        SharedPrefHolder.getInstance(getActivity()).setRingStatus(false);
                    return true;
                }
            });
//
            bindPreferenceSummaryToValue(findPreference(getString(R.string.key_notifications_new_message_ringtone)));

            toggleVibrate.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object o) {
                    if(Boolean.valueOf(o.toString()))
                        SharedPrefHolder.getInstance(getActivity()).setVibrateStatus(true);
                    else
                        SharedPrefHolder.getInstance(getActivity()).setVibrateStatus(false);
                    return true;
                }
            });

            toggleFeedback.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                public boolean onPreferenceClick(Preference preference) {
                    sendFeedback(getActivity());
                    return true;
                }
            });

            try {
                PackageInfo pInfo = getActivity().getPackageManager().getPackageInfo(getActivity().getPackageName(), 0);
                toggleAppVersion.setSummary(pInfo.versionName);
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }

            toggleShare.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    try
                    { Intent i = new Intent(Intent.ACTION_SEND);
                        i.setType("text/plain");
                        i.putExtra(Intent.EXTRA_SUBJECT, "Universal Alarm");
                        String sAux = "\nLet me recommend you this application\n\n";
                        sAux = sAux + "https://play.google.com/store/apps/details?id=" + getActivity().getPackageName();
                        i.putExtra(Intent.EXTRA_TEXT, sAux);
                        startActivity(Intent.createChooser(i, "Share with"));
                    }
                    catch(Exception e)
                    { //e.toString();
                    }
                    return true;

                }
            });

            toggleRate.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Uri uri = Uri.parse("market://details?id=" + getActivity().getPackageName());
                    Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
                    goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY |
                            Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                            Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
                    try {
                        startActivity(goToMarket);
                    } catch (ActivityNotFoundException e) {
                        startActivity(new Intent(Intent.ACTION_VIEW,
                                Uri.parse("http://play.google.com/store/apps/details?id=" + getActivity().getPackageName())));
                    }
                    return true;

                }
            });

            toggleFaq.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    String endpoint=AppConstants.FAQ;
                    Intent browserIntent = new Intent(getActivity(), WebviewActivity.class);
                    browserIntent.putExtra("endpoint",endpoint);
                    startActivity(browserIntent);
                    return true;
                }
            });

            togglePrivacyPolicy.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    String endpoint=AppConstants.PP;
                    Intent browserIntent = new Intent(getActivity(), WebviewActivity.class);
                    browserIntent.putExtra("endpoint",endpoint);
                    startActivity(browserIntent);
                    return true;
                }
            });

            toggleTnc.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    String endpoint=AppConstants.TNC;
                    Intent browserIntent = new Intent(getActivity(), WebviewActivity.class);
                    browserIntent.putExtra("endpoint",endpoint);
                    startActivity(browserIntent);
                    return true;
                }
            });

            toggleTheme.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    new MaterialDialog.Builder(getActivity())
                            .title(R.string.dialogChooseThemeBtTitle)
                            .items(R.array.themes)
                            .itemsCallbackSingleChoice(-1, new MaterialDialog.ListCallbackSingleChoice() {
                                @Override
                                public boolean onSelection(MaterialDialog dialog, View view, int which, CharSequence text) {
                                    SharedPrefHolder.getInstance(getActivity()).setTheme(getResources().getStringArray(R.array.themes)[which]);
                                    toggleTheme.setSummary(getResources().getStringArray(R.array.themes)[which]);
                                    return true;
                                }
                            })
                            .positiveText(R.string.dialogChooseThemeBtText)
                            .show();
                    return true;
                }
            });

        }
    }

    private static void bindPreferenceSummaryToValue(Preference preference) {
        preference.setOnPreferenceChangeListener(sBindPreferenceSummaryToValueListener);

        sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                PreferenceManager
                        .getDefaultSharedPreferences(preference.getContext())
                        .getString(preference.getKey(), ""));
    }

    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
        @Override
        public boolean onPreferenceChange(Preference preference, Object newValue) {
            String stringValue = newValue.toString();
            if (preference instanceof RingtonePreference) {
                if (TextUtils.isEmpty(stringValue)) {
                    preference.setSummary(R.string.pref_ringtone_silent);

                } else {
                    Ringtone ringtone = RingtoneManager.getRingtone(
                            preference.getContext(), Uri.parse(stringValue));
                    SharedPrefHolder.getInstance(preference.getContext()).setRingtoneUri(Uri.parse(stringValue));

                    if (ringtone == null) {

                        preference.setSummary(R.string.summary_choose_ringtone);
                    } else {
                        String name = ringtone.getTitle(preference.getContext());
                        preference.setSummary(name);
                    }
                }
            }
            return true;
        }
    };

    public static void sendFeedback(Context context) {
        String body = null;
        try {
            body = context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName;
            body = "\n\n-----------------------------\nPlease don't remove this information\n Device OS: Android \n Device OS version: " +
                    Build.VERSION.RELEASE + "\n App Version: " + body + "\n Device Brand: " + Build.BRAND +
                    "\n Device Model: " + Build.MODEL + "\n Device Manufacturer: " + Build.MANUFACTURER;
        } catch (PackageManager.NameNotFoundException e) {
        }

        String addresses="mohammadkshf2093@gmail.com";
        Uri uri = Uri.parse("mailto:" + addresses).buildUpon().build();
        Intent emailIntent = new Intent(Intent.ACTION_SENDTO, uri);
        emailIntent.putExtra(Intent.EXTRA_SUBJECT, "Query from Universal Alarm app");
        emailIntent.putExtra(Intent.EXTRA_TEXT, body);
        context.startActivity(emailIntent);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/mdkashif/universalarm/activities/SettingsActivity.kt
package com.mdkashif.universalarm.activities

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.media.RingtoneManager
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.preference.RingtonePreference
import android.text.Spannable
import android.text.SpannableString
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.view.MenuItem
import android.widget.ImageView
import android.widget.TextView
import androidx.preference.PreferenceFragmentCompat
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.afollestad.materialdialogs.list.listItemsSingleChoice
import com.mdkashif.universalarm.R
import com.mdkashif.universalarm.base.BaseActivity
import com.mdkashif.universalarm.persistence.AppPreferences
import com.mdkashif.universalarm.utils.AppConstants
import com.pkmmte.view.CircularImageView
import jp.wasabeef.blurry.Blurry

class SettingsActivity : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        delegate.setLocalNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.title = resources.getString(R.string.settings)
        supportFragmentManager.beginTransaction().replace(android.R.id.content, SettingsFragment()).commit()
    }

    class SettingsFragment : PreferenceFragmentCompat(), androidx.preference.Preference.OnPreferenceClickListener {
        private lateinit var toggleVibrate: androidx.preference.Preference
        private val googlePlayUrl = "http://play.google.com/store/apps/details?id="
        private lateinit var mActivity: SettingsActivity
        private lateinit var mIntent: Intent

        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.fragment_settings, rootKey)

            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleAbout)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleRate)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleShare)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleFaq)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitlePP)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleTNC)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleSendFeedback)).onPreferenceClickListener = this
            findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleSnooze)).onPreferenceClickListener = this

            bindPreferenceSummaryToValue(findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleRingtone)))

            toggleVibrate = findPreference<androidx.preference.Preference>(getString(R.string.prefKeyTitleVibrate))
            toggleVibrate.onPreferenceChangeListener = object : androidx.preference.Preference.OnPreferenceChangeListener {
                override fun onPreferenceChange(preference: androidx.preference.Preference, o: Any): Boolean {
                    AppPreferences.vibrateStatus = java.lang.Boolean.valueOf(o.toString())
                    return true
                }
            }
        }

        private fun showAboutDevDialog() {
            val dialogAboutDev = MaterialDialog(mActivity).show {
                cancelable(true)
                customView(R.layout.layout_about_dev)
            }

            val view = dialogAboutDev.getCustomView()

            val avatar: CircularImageView = view.findViewById(R.id.avatar)
            val bitmap: Bitmap
            val ivDevBlurryImage: ImageView = view.findViewById(R.id.ivDevBlurryImage)
            val ivLinkedIn: ImageView = view.findViewById(R.id.ivLinkedIn)
            val ivGmail: ImageView = view.findViewById(R.id.ivGmail)
            val ivGooglePlay: ImageView = view.findViewById(R.id.ivGooglePlay)
            val ivStackOverFlow: ImageView = view.findViewById(R.id.ivStackOverFlow)
            val ivGithub: ImageView = view.findViewById(R.id.ivGithub)
            val tvMadeWithLove: TextView = view.findViewById(R.id.madeWithLove)
            val tvOpenLicenses: TextView = view.findViewById(R.id.opensource)
            val tvVersion: TextView = view.findViewById(R.id.tvVersion)

            val spannable = SpannableString(getString(R.string.summary_about))
            spannable.setSpan(ForegroundColorSpan(resources.getColor(R.color.red)), 10, 11, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            tvMadeWithLove.setText(spannable, TextView.BufferType.SPANNABLE)

            if (avatar.drawable != null) {
                bitmap = (avatar.drawable as BitmapDrawable).bitmap
                Blurry.with(activity)
                        .radius(5)
                        .sampling(2)
                        .color(Color.argb(100, 0, 0, 0))
                        .async()
                        .animate(1000).from(bitmap).into(ivDevBlurryImage)
            }

            ivLinkedIn.setOnClickListener {
                mIntent = Intent(Intent.ACTION_VIEW, Uri.parse("linkedin://add/%@" + "mdkashif2093"))
                val packageManager = activity!!.packageManager
                val list = packageManager.queryIntentActivities(mIntent, PackageManager.MATCH_DEFAULT_ONLY)
                if (list.isEmpty()) {
                    mIntent = Intent(Intent.ACTION_VIEW, Uri.parse("http://www.linkedin.com/profile/view?id=" + "mdkashif2093"))
                }
                mActivity.executeIntent(mIntent, false)
            }

            ivGmail.setOnClickListener {
                val addresses = "mohammadkshf2093@gmail.com"
                val uri = Uri.parse("mailto:$addresses").buildUpon().build()
                mIntent = Intent(Intent.ACTION_SENDTO, uri)
                mIntent.putExtra(Intent.EXTRA_SUBJECT, "Query!")
                mActivity.executeIntent(mIntent, false)
            }

            ivGooglePlay.setOnClickListener {
                try {
                    mActivity.executeIntent(Intent(Intent.ACTION_VIEW, Uri.parse("market://search?q=pub:Mohammad+Kashif+Khan")), false)
                } catch (anfe: android.content.ActivityNotFoundException) {
                    mActivity.executeIntent(Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/developer?id=Mohammad+Kashif+Khan")), false)
                }
            }

            ivStackOverFlow.setOnClickListener { mActivity.executeIntent(Intent(Intent.ACTION_VIEW, Uri.parse("https://stackoverflow.com/users/5518744/kashif-k")), false) }

            ivGithub.setOnClickListener { mActivity.executeIntent(Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/mohammadkashifkhan")), false) }

            tvOpenLicenses.setOnClickListener {
                MaterialDialog(mActivity).show {
                    cancelable(true)
                    title(R.string.openSourceLicenses)
                    customView(R.layout.layout_open_source_licenses, scrollable = true)
                }
            }

            try {
                val pInfo = activity!!.packageManager.getPackageInfo(activity!!.packageName, 0)
                tvVersion.text = "Version: ${pInfo.versionName}"
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        override fun onPreferenceClick(preference: androidx.preference.Preference?): Boolean {
            when (preference!!.title) {
                getString(R.string.prefKeyTitleSnooze) -> {
                    MaterialDialog(mActivity).show {
                        title(R.string.snoozeTitle)
                        listItemsSingleChoice(R.array.snoozeTimings) { dialog, index, text ->
                            AppPreferences.snoozeTimeArrayPosition = index
                        }
                    }
                }

                getString(R.string.prefKeyTitleAbout) ->
                    showAboutDevDialog()

                getString(R.string.prefKeyTitleRate) -> {
                    val uri = Uri.parse("market://details?id=" + activity!!.packageName)
                    mIntent = Intent(Intent.ACTION_VIEW, uri)
                    mIntent.addFlags((Intent.FLAG_ACTIVITY_NO_HISTORY or
                            Intent.FLAG_ACTIVITY_NEW_DOCUMENT or
                            Intent.FLAG_ACTIVITY_MULTIPLE_TASK))
                    try {
                        mActivity.executeIntent(mIntent, false)
                    } catch (e: ActivityNotFoundException) {
                        mActivity.executeIntent(Intent(Intent.ACTION_VIEW,
                                Uri.parse(googlePlayUrl + activity!!.packageName)), false)
                    }
                }

                getString(R.string.prefKeyTitleShare) -> {
                    try {
                        mIntent = Intent(Intent.ACTION_SEND)
                        mIntent.type = "text/plain"
                        mIntent.putExtra(Intent.EXTRA_SUBJECT, "Universal Alarm")
                        var sAux = "\nLet me recommend you this application\n\n"
                        sAux = sAux + googlePlayUrl + activity!!.packageName
                        mIntent.putExtra(Intent.EXTRA_TEXT, sAux)
                        mActivity.executeIntent(Intent.createChooser(mIntent, "Share with"), false)
                    } catch (e: Exception) { //e.toString();
                    }
                }

                getString(R.string.prefKeyTitleFaq) -> {
                    mIntent = Intent(activity, WebviewActivity::class.java)
                    mIntent.putExtra("endpoint", AppConstants.FAQ)
                    mActivity.executeIntent(mIntent, false)
                }

                getString(R.string.prefKeyTitlePP) -> {
                    mIntent = Intent(activity, WebviewActivity::class.java)
                    mIntent.putExtra("endpoint", AppConstants.PP)
                    mActivity.executeIntent(mIntent, false)
                }

                getString(R.string.prefKeyTitleTNC) -> {
                    mIntent = Intent(activity, WebviewActivity::class.java)
                    mIntent.putExtra("endpoint", AppConstants.TNC)
                    mActivity.executeIntent(mIntent, false)
                }

                getString(R.string.prefKeyTitleSendFeedback) ->
                    mActivity.sendFeedback()
            }
            return true
        }

        override fun onAttach(context: Context) {
            super.onAttach(context)
            mActivity = context as SettingsActivity
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private fun bindPreferenceSummaryToValue(preference: androidx.preference.Preference) {
            preference.onPreferenceChangeListener = sBindPreferenceSummaryToValueListener

            sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                    PreferenceManager
                            .getDefaultSharedPreferences(preference.context)
                            .getString(preference.key, ""))
        }

        private val sBindPreferenceSummaryToValueListener = object : androidx.preference.Preference.OnPreferenceChangeListener {
            override fun onPreferenceChange(preference: androidx.preference.Preference, newValue: Any): Boolean {
                val stringValue = newValue.toString()
                if (preference is RingtonePreference) {
                    if (TextUtils.isEmpty(stringValue)) {
                        preference.setSummary(R.string.pref_ringtone_silent)
                    } else {
                        val ringtone = RingtoneManager.getRingtone(
                                preference.getContext(), Uri.parse(stringValue))
                        AppPreferences.ringtoneUri = Uri.parse(stringValue).toString()

                        if (ringtone == null)
                            preference.setSummary(R.string.prefSummaryRingtone)
                        else {
                            val name = ringtone.getTitle(preference.getContext())
                            preference.setSummary(name)
                        }
                    }
                }
                return true
            }
        }
    }
}
